li-protobuf/src/main/java/com/linkedin/data/protobuf/InputStreamReader.java

```diff
@@ -0,0 +1,655 @@
+  {
+    final int size = readInt32();
+    if (size > 0)
+    {
+      // If we can fit into a buffer, read directly off the buffer,
+      if (size < _bufferSize)
+      {
+        // Slow path: We can fit into a buffer, but there aren't enough bytes available in the current buffer.
+        // Refill!
+        if (size > (_bufferSize - _pos))
+        {
+          refillBuffer(size);
+        }
+
+        String value = Utf8Utils.decode(_buffer, _pos, size, _textBuffer);
+        _pos += size;
+        return value;
+      }
+      else
+      {
+        Utf8Utils.LongDecoderState state = new InputStreamLongDecoderState(_buffer, _pos, _bufferSize, _input);
+        String value = Utf8Utils.decodeLong(state, size, _textBuffer);
+        _pos = state.getPosition();
+        _bufferSize = state.getBufferSize();
+        return value;
+      }
+    }
+    else if (size == 0)
+    {
+      return "";
+    }
+    else
+    {
+      throw new IOException("Read negative size: " + size + ". Invalid string");
+    }
+  }
+
+  @Override
+  public byte[] readByteArray() throws IOException
+  {
+    final int size = readInt32();
+    if (size <= (_bufferSize - _pos) && size > 0)
+    {
+      // Fast path: We already have the bytes in a contiguous _buffer, so
+      // just copy directly from it.
+      final byte[] result = Arrays.copyOfRange(_buffer, _pos, _pos + size);
+      _pos += size;
+      return result;
+    }
+    else
+    {
+      // Slow path: Build a byte array first then copy it.
+      return readRawBytesSlowPath(size);
+    }
+  }
+
+  @Override
+  public int readInt32() throws IOException
+  {
+    // See implementation notes for readInt64
+    fastpath:
+    {
+      int tempPos = _pos;
+
+      if (_bufferSize == tempPos)
+      {
+        break fastpath;
+      }
+
+      final byte[] buffer = this._buffer;
+      int x;
+      if ((x = buffer[tempPos++]) >= 0)
+      {
+        _pos = tempPos;
+        return x;
+      }
+      else if (_bufferSize - tempPos < 9)
+      {
+        break fastpath;
+      }
+      else if ((x ^= (buffer[tempPos++] << 7)) < 0)
+      {
+        x ^= (~0 << 7);
+      }
+      else if ((x ^= (buffer[tempPos++] << 14)) >= 0)
+      {
+        x ^= (~0 << 7) ^ (~0 << 14);
+      }
+      else if ((x ^= (buffer[tempPos++] << 21)) < 0)
+      {
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21);
+      }
+      else
+      {
+        int y = buffer[tempPos++];
+        x ^= y << 28;
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21) ^ (~0 << 28);
+        if (y < 0
+            && buffer[tempPos++] < 0
+            && buffer[tempPos++] < 0
+            && buffer[tempPos++] < 0
+            && buffer[tempPos++] < 0
+            && buffer[tempPos++] < 0)
+        {
+          break fastpath; // Will throw malformedVarint()
+        }
+      }
+      _pos = tempPos;
+      return x;
+    }
+    return (int) readRawVarint64SlowPath();
+  }
+
+  @Override
+  public long readInt64() throws IOException
```

Comment:
Is this the same or similar code as that one in `ByteArrayReader`? Is it possible to share some code.

Reply:
Looks similar but there are subtle differences in between the method in several places. Choose to keep it separate for better clarity/readability, especially given that this method will hardly ever change.

Google's protobuf implementation does the same FWIW